### ISSUE TITLE
For symmetry, define a logicalLocation object 

### ISSUE THREAD

**Issue by michaelcfanning** [oasis-tcs/sarif-spec#145]
so that we have location.physcialLocation and location.logicalLocation. The new type would encapsulate the following. The existing 'logicalLocation' object is actually a logical location sub-element in most cases (not so for a root object, like a namespace, however). We could call these logicalLocationElements.

        "fullyQualifiedLogicalName": {
          "description": "The human-readable fully qualified name of the logical location where the analysis tool produced the result. If 'logicalLocationKey' is not specified, this member is can used to retrieve the location logicalLocation from the logicalLocations dictionary, if one exists.",
          "type": "string"
        },

        "logicalLocationKey": {
          "description": "A key used to retrieve the location logicalLocation from the logicalLocations dictionary, when the string specified by 'fullyQualifiedLogicalName' is not unique.",
          "type": "string"
        },

        "decoratedName": {
          "description": "The machine-readable fully qualified name for the logical location where the analysis tool produced the result, such as the mangled function name provided by a C++ compiler that encodes calling convention, return type and other details along with the function name.",
          "type": "string"
        },

**Comment by lgolding** [comment]
@michaelcfanning Yes, we've discussed this one. Now that the rest of the location redesign is done, this is pretty straightforward. I'm happy to take this into CSD.1

**Comment by lgolding** [comment]
@michaelcfanning Ok, working on this now. What do you think of `logicalLocationSegment` for the elements of the `file.logicalLocations` dictionary?

**Comment by lgolding** [comment]
No, I take it back. they aren't segments. They can be an entire name like `namespaceA::namespaceB::classC::methodD`. The thing that `run.logicalLocation` gives you is nesting information (by means of the `parentKey` property), without you having to know how to parse the FQLN. So, ok, "segment" is wrong; I'm not sure "element" is better. I need to think some more. I'll use a placeholder name while I do the writing.

**Comment by lgolding** [comment]
@michaelcfanning  Let's consider doing this a different way. I understand that our original motivations for this change were:

1. Introduce a symmetry between the physical and logical aspects of the location object:
```
location
  physicalLocation
  logicalLocation
```
2. DRY out the repeated occurrences of the property pair (`fullyQualifiedLogicalName`, `logicalLocationKey`)

First of all, notice that since the location redesign (#130), that property pair occurs in exactly one place: in the location object. There's nothing to DRY out.

Second, the existing `logicalLocation` objects in the `run.logicalLocations` dictionary already contains some descriptive information about the logical location, namely the `name` and `kind` properties. That suggests that `decoratedName` should also live in the dictionary.

I propose:

- Move `decoratedName` to the existing `logicalLocation` object.
- Move `fullyQualifiedLogicalName` to the existing `logicalLocation` object.
- Leave behind only `logicalLocationKey` (which is usually equal to `fullyQualifiedLogicalName`) in the `location` object.

Now, this is not fully symmetric between physical and logical locations. But in fact, our design is not fully symmetric. We don't have a `physicalLocations` dictionary (although we have discussed it);
rather, fully populated `physicalLocation` objects live throughout the `result` object (and elsewhere). But we _do_ have a `logicalLocations` dictionary, which is a sensible place to hold information like `decoratedName`. So it makes sense for `location` to have a `physicalLocation` property but a `logicalLocationKey` (emphasis on "key") property.

**Comment by michaelcfanning** [comment]
Suggest leaving behind 'fullyQualifiedName' and making it clear that if this is used as a key into the logical locations table, the actually FQDN may have been redefined (because the key was revised to provide uniqueness).

And we should play the same trick with ruleId + ruleKey elsewhere.

**Comment by lgolding** [comment]
@michaelcfanning We don't need `logicalLocation.decoratedName`. Look what the spec says about `logicalLocation.name`:

> A `logicalLocation` object **SHALL** contain a property named `name` whose value is a string that identifies the construct in which the result occurred. For example, this property might contain the name of a class or a method.
> The `name` property does need to be suitable for display.
>
>  EXAMPLE 1: A C++ analysis tool might emit the name property of a function as the “decorated” function name, which encodes the function signature in a manner that is compiler-dependent and not easily readable.

We just need to add `fullyQualifiedLogicalName` -- and that only if it differs from the dictionary key.

**Comment by lgolding** [comment]
@michaelcfanning And I suggest we name it `logicalLocation.fullyQualifiedName` (not `fullyQualifiedLogicalName`). That is:
```
logicalLocation
  name
  fullyQualifiedName
  ...

result:
  fullyQualifiedLogicalName

**Comment by michaelcfanning** [comment]
No we require decorated name, it is a unique non human readable thing that can’t be parsed into constituent elements. Search web for examples.

The spec already makes this clear, name should be readable but decorated name is not.

Btw we need to add fullyQualifiedName to logical location in case it needs to be redefined (if it isn’t already there)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by lgolding** [comment]
@michaelcfanning And a final reconsideration. It's true that the spec says `logicalLocation.name` needn't be readable. But when I tried it out in the comprehensive example in the spec, it just looked wrong to me:
```
# BEFORE
"logicalLocations": {
  "collections::list::add": {
    "name": "add",
    "kind": "function",
    "parentKey": "collections::list"
  },
  ...

# AFTER
"logicalLocations": {
  "collections::list::add": {
    "name": "?add@list@collections@@QAEXH@Zadd",    # ICK! Where can I see just plain "add"?
    "kind": "function",
    "parentKey": "collections::list"
  },
  ...
```

I propose to change the spec so that `name` is readable, and you use `decoratedName` for the decorated name:
```
# PROPOSED
"logicalLocations": {
  "collections::list::add": {
    "name": "add",
    "decoratedName": "?add@list@collections@@QAEXH@Zadd",
    "kind": "function",
    "parentKey": "collections::list"
  },
  ...
```

This won't be controversial: it's what we both imagined we were proposing before I realized that as the spec stood, `name` didn't have to be readable. I'll fix that in the spec.

In short, we have:
```
logicalLocation
  name
  decoratedName
  fullyQualifiedName
  ...

location:
  fullyQualifiedLogicalName
```
